### ISSUE TITLE
Single instance only

### ISSUE THREAD

**Issue by viencokhi** [BishopFox/sliver#231]
We should implement some kind of mutex so only one instance of implant running at a time.

**Comment by rkervella** [comment]
Can you define a specific use case?

**Comment by moloch--** [comment]
Seems like a good idea to me, you may run into situations where your persistence technique results in multiple executions of a payload based on user activity and it would probably be a good idea to limit the number of sessions that get created.

**Comment by rkervella** [comment]
Makes sense. I'll play with `CreateMutexA` and see what's the most convenient way to integrate that. We'll probably need to modify the `generate` command to have a common mutex name per implant build (or we could use the implant name as a mutex name too).

**Comment by moloch--** [comment]
Yea we'll want some way of doing this on MacOS and Linux too methinks. We'll want to consider if this gets enabled by default.

**Comment by rkervella** [comment]
Good point.

**Comment by moloch--** [comment]
It may also be nice to have a scheme were we can limit execution `n` instances.

**Comment by rkervella** [comment]
So basically a semaphore.

**Comment by rkervella** [comment]
Looks like we have a Linux compatible sample here, need to see what's needed for unix / OS X to make it work too: https://github.com/shubhros/drunkendeluge/blob/master/semaphore/semaphore.go

**Comment by moloch--** [comment]
Cool, let's make this per-generated-implant too, just randomize all the values.

**Comment by moloch--** [comment]
Probably implement this in the existing `limits` package too.

**Comment by Matir** [comment]
Just a thought: should they be per-user on the host?  So if you have an unprivileged sliver and then privesc, launch a 2nd sliver, it would be nice to not have this interfere.

**Comment by usiegl00** [comment]
I have some c / go code that will get a UUID per host.
The server could use this uuid to decide whether to accept the connection.

Ex. If there is already a session with the same UID, GID and UUID then refuse / terminate the connection.

Linux:
```go
package main

import (
	"fmt"
	"io/ioutil"
)

func main() {
	uuid, err := ioutil.ReadFile("/var/lib/dbus/machine-id")
	if err != nil {
		panic(err)
	}
	fmt.Printf("%s-%s-%s-%s-%s\n",
		uuid[0:8],
		uuid[8:12], uuid[12:16], uuid[16:20],
		uuid[20:32])
}

```
Darwin:
```c
#include <unistd.h>
#include <uuid/uuid.h>
#include <stdio.h>

int main() {
  uuid_t id;
  const struct timespec wait = {5, 0};
  if (gethostuuid(id, &wait) == 0) {
    char out[36];
    uuid_unparse_lower(id, out);
    printf("%s\n", out);
  };
  return 0;
};
```
Windows:
```go
package main

import (
	"bytes"
	"fmt"
	"os/exec"
)

func main() {
	cmd := exec.Command("wmic.exe", "csproduct", "get", "UUID")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		panic(err)
	}
	str := out.String()
	fmt.Printf("String: %s\n", str[41:77])
}
```

**Comment by usiegl00** [comment]
I have a UUID implementation in progress, but am running into a road block with dependencies.
I was able to add my new sliver code to the srcfiles. (server/assets.go)
However I am unable to add the "github.com/google/uuid" package to the sliver build.
When generating a sliver I get the following error in the log:
```text
s97.go:49:2: cannot find package "github.com/google/uuid" in any of:
	/Users/iron/.sliver/go/src/github.com/google/uuid (from $GOROOT)
	/Users/iron/.sliver/slivers/darwin/amd64/OK_COMPLEMENT/obfuscated/src/github.com/google/uuid (from $GOPATH)
```
If I could have some pointers on how to add the "uuid" package to the sliver build that would be fantastic.

**Comment by rkervella** [comment]
@usiegl00 I'm not found of this solution two reasons:

- we want to avoid the implant running multiple time on the same host, and that should be handled implant-side. Using mutexes / semaphores is what makes sense here.
- there's an edge case where the same host can have multiple UUIDs, which would make this solution unreliable for this case

**Comment by usiegl00** [comment]
Makes Sense.

UUIDs work well for persistence however, allowing the server to match based on UID in the edge cases.
